packages/flutter_web_plugins/lib/src/history/url_strategy.dart

```diff
@@ -0,0 +1,332 @@
+///
+/// In order to use this [UrlStrategy] for an app, it needs to be set like this:
+///
+/// ```dart
+/// import 'package:flutter_web_plugins/flutter_web_plugins.dart';
+///
+/// // Somewhere before calling `runApp()` do:
+/// setUrlStrategy(PathUrlStrategy());
+/// ```
+class PathUrlStrategy extends UrlStrategy {
+  /// Creates an instance of [PathUrlStrategy].
+  ///
+  /// The [PlatformLocation] parameter is useful for testing to avoid
+  /// interacting with the actual browser.
+  PathUrlStrategy([
+    this._platformLocation = const BrowserPlatformLocation(),
+  ]) : _basePath = stripTrailingSlash(extractPathname(
+          _platformLocation.getBaseHref(),
+        ));
+
+  final PlatformLocation _platformLocation;
+  final String _basePath;
+
+  @override
+  ui.VoidCallback onPopState(html.EventListener fn) {
+    _platformLocation.onPopState(fn);
+    return () => _platformLocation.offPopState(fn);
+  }
+
+  @override
+  String getPath() {
+    final String path = _platformLocation.pathname + _platformLocation.search;
+    if (_basePath.isNotEmpty && path.startsWith(_basePath)) {
+      return ensureLeadingSlash(path.substring(_basePath.length));
+    }
+    return ensureLeadingSlash(path);
+  }
+
+  @override
+  dynamic getState() => _platformLocation.state;
+
+  @override
+  String prepareExternalUrl(String internalUrl) {
+    if (internalUrl.isNotEmpty && !internalUrl.startsWith('/')) {
+      internalUrl = '/$internalUrl';
+    }
+    return '$_basePath$internalUrl';
+  }
+
+  @override
+  void pushState(dynamic state, String title, String url) {
+    _platformLocation.pushState(state, title, prepareExternalUrl(url));
+  }
+
+  @override
+  void replaceState(dynamic state, String title, String url) {
+    _platformLocation.replaceState(state, title, prepareExternalUrl(url));
+  }
+
+  @override
+  Future<void> go(int count) {
+    _platformLocation.go(count);
+    return _waitForPopState();
+  }
+
+  /// Waits until the next popstate event is fired.
+  ///
+  /// This is useful for example to wait until the browser has handled the
+  /// `history.back` transition.
+  Future<void> _waitForPopState() {
+    final Completer<void> completer = Completer<void>();
+    ui.VoidCallback unsubscribe;
+    unsubscribe = onPopState((_) {
+      unsubscribe();
+      completer.complete();
+    });
+    return completer.future;
+  }
+}
+
+/// [PlatformLocation] encapsulates all calls to DOM apis, which allows the
+/// [UrlStrategy] classes to be platform agnostic and testable.
+///
+/// The [PlatformLocation] class is used directly by all implementations of
+/// [UrlStrategy] when they need to interact with the DOM apis like
+/// pushState, popState, etc...
+abstract class PlatformLocation {
+  ///
+  const PlatformLocation();
+
+  /// Registers an event listener for the `onpopstate` event.
+  ///
+  /// See: https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onpopstate
+  void onPopState(html.EventListener fn);
```

Comment:
Since this is part of the public API we should probably be more Darty and name it `addPopStateListener`?